### ISSUE TITLE
Switch from GMime parser

### ISSUE THREAD

**Issue by vstakhov** [vstakhov/rspamd#1225]
We need a parser that can work with broken messages and extract more information from them.

**Comment by vstakhov** [comment]
Done.

**Comment by jstedfast** [comment]
Was GMime having problems parsing some messages? Would you mind sending me some examples? I'd like to improve GMime's parser to handle them.

Thanks.

**Comment by vstakhov** [comment]
@jstedfast, I would like to say thank you first of all for your GMime project: Rspamd used it for a very long time and I was very pleased to use it in principle.

However, the main problem was that Rspamd required very precise control on MIME parsing as it had to deal with broken messages not for displaying purposes but for extracting data from them. It has some simplifications and complications comparing to a generic MIME parser, such as GMime: for example, I do not need to support streaming mode (which I believe to be the main complication when building parsers) but I have to deal with many non-standard messages that are *intended* to be parsed incorrectly by some adversary side, e.g. spammers. The current architecture is described here: https://gist.github.com/vstakhov/937f253d5935ee4158688932589b1dcc

Here is an incomplete list of messages that were incorrectly handled by GMime:

1) A messages with redundant `Content-Type` headers:

```
Content-Type: text/plain
Content-Type: multipart/alternative
```

Currently, Rspamd always prefers multipart types over plain types and text types unless there is not specific binary type (e.g. if there are `text/plain` and `application/octet-stream`)

2) Messages with broken multipart structures:
  - new parts after closing boundary (e.g. attachment in multipart/mixed after the closing part)
  - incorrect inheritance
  - incorrect multipart type (now Rspamd just ignores the exact `multipart/*` type)

3) Filenames are badly encoded (non-utf8)
4) Incorrect `Content-Transfer-Encoding` (now heuristic based):
  - `8bit` when content is Base64
  - `base64` or `qp` when content is 8bit
5) Bad `content-type`, e.g. `text`

There were many other issues but I cannot find them in my memory. I'll try to find some examples of the issues listed but I'm not sure that I still have them.

**Comment by vstakhov** [comment]
Also: messages with no headers and messages with no body at all, messages with headers and no CRLF at the end.

**Comment by jstedfast** [comment]
@vstakhov 

Wow, thanks so much for your thoughtful response and kind words. I had not considered some of the things you mentioned but now that I have read your response, I can see why you've moved on to using your own custom parser and from a technical standpoint, I agree that it makes the best sense.

Please don't feel like you have to go out of your way to find examples if you don't already have a stash of them handy. I can work with the information you've already given me. It's far more important to me that you spend your free time continuing to make your project awesome :)

Thanks again.

**Comment by vstakhov** [comment]
One more thing: I have to implement my own headers folding algorithm as gmime wasn't able to fold large DKIM signatures correctly. I'm not 100% sure about what was the exact problem but it led to a completely broken header when I tried to fold it. Here is the current algorithm I've designed: https://github.com/vstakhov/rspamd/blob/master/src/libutil/str_util.c#L957 Another important difference is that Rspamd now tries to figure which line endings are mostly used in the headers use this specific line endings and not `\r\n` all the time.

**Comment by vstakhov** [comment]
One more thing I've found about bad messages and GMime: if there are multiple line endings in a message, e.g. some headers are terminated with `\r` and some headers are terminated with `\n` and, finally, some headers are terminated with `\r\n` then GMime failed to detect end-of-headers position and parse headers. Body with mixed line endings was also incorrectly parsed (in that case, body was initially splitted by `\r` symbols and, subsequently, by `\r\n` characters). Currently, I detect the end-of-headers position with this code: https://github.com/vstakhov/rspamd/blob/master/src/libutil/str_util.c#L1422

**Comment by jstedfast** [comment]
Thanks for the info. One of the problems I know GMime used to have until recently was that adding or modifying any parsed header of a GMimeMessage would cause all headers to be re-folded.

That was a pain-point for a lot of apps using GMime, so that was one of the main things I focused on fixing recently.

Not sure if that's what you meant about the folding of DKIM signature headers.

I'll take a closer look at your folding algorithm when I get a chance. From a quick glance, it looks like your approach was to more-or-less tokenize the input as words in order to properly fold them and that is similar to what I did in my MimeKit library (a sort of re-write of GMime in C#). It's something I'd like to do for GMime as well.

Thanks for the info, this has all been helpful.
